File: Sources/SiftCore/Runtime/AppPaths.swift
import Foundation

/// Where Sift reads transcripts from and writes its databases to.
///
/// Both are overridable by environment variable so a run can be pointed at throwaway data:
/// that is how the screenshots in the README are made without putting anyone's real session
/// titles on the internet, and how you can try a build without it touching your live index.
///
///     SIFT_PROJECTS_ROOT=/tmp/demo/projects \
///     SIFT_SUPPORT_DIR=/tmp/demo/support \
///     /Applications/Sift.app/Contents/MacOS/Sift
public enum AppPaths {
    public static let projectsRootKey = "SIFT_PROJECTS_ROOT"
    public static let supportDirectoryKey = "SIFT_SUPPORT_DIR"

    public static var projectsRoot: URL {
        resolve(projectsRootKey) ?? URL(fileURLWithPath: NSHomeDirectory())
            .appendingPathComponent(".claude/projects")
    }

    public static var supportDirectory: URL {
        if let override = resolve(supportDirectoryKey) { return override }
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? URL(fileURLWithPath: NSHomeDirectory()).appendingPathComponent("Library/Application Support")
        return base.appendingPathComponent("Sift", isDirectory: true)
    }

    /// True when either path is redirected, so the app can say so instead of leaving someone
    /// wondering why their sessions are missing.
    public static var isUsingOverriddenPaths: Bool {
        resolve(projectsRootKey) != nil || resolve(supportDirectoryKey) != nil
    }

    static func resolve(_ key: String, environment: [String: String] = ProcessInfo.processInfo.environment) -> URL? {
        guard let raw = environment[key]?.trimmingCharacters(in: .whitespaces), !raw.isEmpty else { return nil }
        return URL(fileURLWithPath: (raw as NSString).expandingTildeInPath)
    }
}

File: windows-winui/Sift.Core/AppPaths.cs
using System.IO;

namespace SiftWinUI.Core;

/// Where Sift reads transcripts from and writes its index to. Both are overridable by
/// environment variable so a run can be pointed at throwaway data, which is how the
/// screenshots are made without putting anyone's real session titles on the internet.
public static class AppPaths
{
    public const string ProjectsRootKey = "SIFT_PROJECTS_ROOT";
    public const string SupportDirKey = "SIFT_SUPPORT_DIR";

    public static string ProjectsRoot => Resolve(ProjectsRootKey)
        ?? Path.Combine(Home, ".claude", "projects");

    public static string SupportDir => Resolve(SupportDirKey)
        ?? Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "Sift");

    public static string IndexPath => Path.Combine(SupportDir, "index.sqlite");

    public static string ClaudeCommand =>
        Environment.GetEnvironmentVariable("SIFT_CLAUDE") is { Length: > 0 } c ? c : "claude";

    private static string Home => Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);

    public static string? Resolve(string key, IDictionary<string, string>? env = null)
    {
        var raw = env is null
            ? Environment.GetEnvironmentVariable(key)
            : (env.TryGetValue(key, out var v) ? v : null);
        raw = raw?.Trim();
        return string.IsNullOrEmpty(raw) ? null : Environment.ExpandEnvironmentVariables(raw);
    }
}

/// Indexes every transcript that changed since last time. Unchanged files are skipped on
/// (size, mtime), so a rescan over thousands of sessions costs almost nothing.
public static class Indexer
{
    public sealed record Result(int Total, int Indexed, int Removed);

    public static Result Reindex(IndexStore store, string projectsRoot, Action<int, int>? progress = null)
    {
        var found = Scanner.ListTranscripts(projectsRoot);
        var known = store.Fingerprints();
        var indexed = 0;

        for (var i = 0; i < found.Count; i++)
        {
            var file = found[i];
            if (known.TryGetValue(file.FilePath, out var seen) &&
                seen.Size == file.Size && seen.Mtime == file.Mtime) continue;
            try
            {
                var row = Scanner.ParseTranscript(file.FilePath);
                row.ProjectId = file.ProjectId;
                row.FileSize = file.Size;
                row.FileMtime = file.Mtime;
                store.Upsert(row);
                indexed++;
            }
            catch (IOException) { /* a transcript being written right now */ }
            if (i % 100 == 0) progress?.Invoke(i, found.Count);
        }

        var removed = store.RemoveMissing(found.Select(f => f.FilePath));
        return new Result(found.Count, indexed, removed);
    }
}
